### ISSUE TITLE
Dubious borrowck warning in 2018 edition

### ISSUE THREAD

**Issue by vbarrielle** [rust-lang/rust#59645]
When converting my project to the 2018 edition, I've had a borrowck warning [on an innocent looking line](https://github.com/vbarrielle/sprs/blob/908e0f0a13309fdc15d122e21cf8f7b07245ab2b/src/sparse/triplet_iter.rs#L122).

I've reproduced the issue with a minimal example:

```rust
fn foo<T: Copy + std::ops::AddAssign>(v: &mut Vec<T>, i: usize) {
    v[i] += v[i-1];
}

fn bar(v: &mut Vec<usize>, i: usize) {
    v[i] += v[i-1];
}

fn main() {
    let mut a = vec![0,1,2,3];
    
    foo(&mut a, 2);
    bar(&mut a, 2);
    
    println!("{:?}", a);
}
```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=83e78452454f75ddd4205cc8fb3eafa3))

Output:

```
[0, 1, 4, 3]

```

Errors:

```
   Compiling playground v0.0.1 (/playground)
warning[E0502]: cannot borrow `*v` as immutable because it is also borrowed as mutable
 --> src/main.rs:2:13
  |
2 |     v[i] += v[i-1];
  |     --------^-----
  |     |       |
  |     |       immutable borrow occurs here
  |     mutable borrow occurs here
  |     mutable borrow later used here
  |
  = warning: this error has been downgraded to a warning for backwards compatibility with previous releases
  = warning: this represents potential undefined behavior in your code and this warning will become a hard error in the future

    Finished dev [unoptimized + debuginfo] target(s) in 0.92s
     Running `target/debug/playground`

```

There is no warning generated if using a `usize`, but replacing it with a type `T: Copy + std::ops::AddAssign` triggers the warning. However in my opinion if this is safe for a `usize` this should also be safe for any such `T`. Compiling this example on the 2015 edition runs fine, so it might be related to NLL (I think it's still not turned on in the 2015 edition?).

**Comment by matthewjasper** [comment]
The warning is correct. Overloaded compound assignment operators unfortunately evaluate their operands in the opposite order to the built in ones.

**Comment by vbarrielle** [comment]
Thanks for the explanation. I guess I can close it then.
